pallets/file-system/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+        ///
+        /// The transaction will fail if the XOR between the file ID and the BSP ID is not below the threshold,
+        /// so a BSP is strongly advised to check beforehand. Another reason for failure is
+        /// if the maximum number of BSPs has been reached. A successful assignment as BSP means
+        /// that some of the collateral tokens of that MSP are frozen.
+        #[pallet::call_index(4)]
+        #[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1).ref_time())]
+        pub fn bsp_volunteer(
+            origin: OriginFor<T>,
+            location: FileLocation<T>,
+            fingerprint: Fingerprint<T>,
+            bsp_multiaddress: MultiAddress<T>,
+        ) -> DispatchResult {
+            // Check that the extrinsic was signed and get the signer.
+            let who = ensure_signed(origin)?;
+
+            // Perform validations and register Storage Provider as BSP for file.
+            Self::do_bsp_volunteer(who.clone(), location.clone(), fingerprint)?;
+
+            // Emit new BSP volunteer event.
+            Self::deposit_event(Event::AcceptedBspVolunteer {
+                who,
+                location,
+                fingerprint,
+                bsp_multiaddress,
+            });
+
+            Ok(())
+        }
+
+        /// Executed by a BSP to stop storing a file.
+        ///
+        /// A compensation should be provided for the user, to deter this behaviour. A successful execution of this extrinsic automatically generates a storage request for that file with one remaining_bsps_slot left, and if a storage request for that file already exists, the slots left are incremented in one. It also automatically registers a challenge for this file, for the next round of storage proofs, so that the other BSPs and MSP who are storing it would be forced to disclose themselves then.
+        #[pallet::call_index(5)]
+        #[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1).ref_time())]
+        pub fn bsp_stop_storing(_origin: OriginFor<T>) -> DispatchResult {
+            todo!()
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+    where
+        u32: TryFrom<BlockNumberFor<T>>,
+    {
+        fn on_idle(block: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+            let db_weight = T::DbWeight::get();
+
+            // Return early if the remaining weight is not enough to perform the operation
+            if !remaining_weight.all_gte(
+                db_weight.reads_writes(
+                    1,
+                    T::MaxExpiredStorageRequests::get()
+                        .try_into()
+                        .expect("u32 would fit into u64"),
+                ),
+            ) {
+                return Weight::zero();
+            }
+
+            let mut used_weight = db_weight.reads(1);
+
+            let mut expired_requests = match StorageRequestExpirations::<T>::take(&block) {
```

Comment:
Is it possible that this doesn't get executed for block.

Reply:
I am not sure I understand, can you clarify

Reply:
Yes sorry. I meant to say that there should be a way to keep track of the last block for which this `on_idle` was executed, and effectively cleared storage. This is in case that in some heavily used block, `on_idle` doesn't get to be executed, so it lags behind.

Reply:
Put some TODOs for these scenarios.